micro-rdk/src/common/conn/viam.rs

```diff
@@ -0,0 +1,756 @@
+        let robot = Arc::new(Mutex::new(robot));
+
+        let app_client = self.connect_to_app().await.unwrap();
+
+        let certs = app_client.get_certificates().await.unwrap();
+
+        if let HTTP2Server::HTTP2Connector(s) = &mut self.http2_server {
+            s.set_server_certificates(
+                certs.tls_certificate.into_bytes(),
+                certs.tls_private_key.into_bytes(),
+            );
+        }
+        let ip = self.network.get_ip();
+
+        let config = app_client.get_app_config(Some(ip)).await.unwrap();
+        self.rpc_host = config.0.config.unwrap().cloud.unwrap().fqdn;
+
+        let (tx, rx) = async_channel::bounded(1);
+        let inner = RobotServer::new(
+            &self.http2_server,
+            &self.webrtc_configuration,
+            self.executor.clone(),
+            robot.clone(),
+            &self.mdns,
+            self.http2_server_port,
+            rx,
+            &self.network,
+        );
+
+        self.app_client_tasks.push(Box::new(SignalingTask {
+            sender: tx,
+            rpc_host: self.rpc_host.clone(),
+        }));
+        let mut tasks: FuturesUnordered<_> = FuturesUnordered::new();
+        tasks.push(Either::Right(self.run_app_client_tasks(Some(app_client))));
+        tasks.push(Either::Left(inner.run()));
+        while let Some(_) = tasks.next().await {}
+    }
+    async fn connect_to_app(&self) -> Result<AppClient, AppClientError> {
+        //ugly hack to remove last /
+        //needs to change that but would need to update GrpcClient
+        let uri = self.app_uri.to_string();
+        let mut chars = uri.chars();
+        chars.next_back();
+        let uri = chars.as_str();
+        let robot_creds = self.storage.get_robot_credentials().unwrap();
+        let app_client_io = self
+            .http2_connector
+            .connect_to(&self.app_uri)
+            .map_err(AppClientError::AppClientIoError)?
+            .await
+            .map_err(AppClientError::AppClientIoError)?;
+        let grpc_client = GrpcClient::new(app_client_io, self.executor.clone(), uri)
+            .await
+            .map_err(AppClientError::AppGrpcClientError)?;
+
+        AppClientBuilder::new(Box::new(grpc_client), robot_creds.clone())
+            .build()
+            .await
+    }
+    async fn run_app_client_tasks(&self, app_client: Option<AppClient>) {
+        let app_client = if let Some(app) = app_client {
+            app
+        } else {
+            self.connect_to_app().await.unwrap()
+        };
+        let mut app_client_tasks: FuturesUnordered<AppClientTaskRunner> = FuturesUnordered::new();
+        for task in &self.app_client_tasks {
+            app_client_tasks.push(AppClientTaskRunner {
+                app_client: &app_client,
+                invoker: task,
+                timer: None,
+                task: None,
+            });
+        }
+        while let Some(res) = app_client_tasks.next().await {
+            res.unwrap();
```

Comment:
It's more verbose, but I think it would be clearer to do an `if res.is_err() { panic!(...); }`. I think unwrap obfuscates the fact that when an app client task fails we want to cause a restart.

Reply:
This where we would implement the code handling the restarting of  app_client and so on. I haven't added the logic around this yet

Reply:
ah ok, makes sense